processor/spanmetricsprocessor/processor.go

```diff
@@ -141,15 +198,196 @@ func (p *processorImp) ConsumeTraces(ctx context.Context, traces pdata.Traces) e
+		dpCalls.SetStartTime(pdata.TimeToUnixNano(p.startTime))
+		dpCalls.SetTimestamp(pdata.TimeToUnixNano(time.Now()))
+		dpCalls.SetValue(p.callSum[key])
+
+		dpCalls.LabelsMap().InitFromMap(p.metricKeyToDimensions[key])
+
+		mCalls := pdata.NewMetric()
+		mCalls.SetDataType(pdata.MetricDataTypeIntSum)
+		mCalls.SetName("calls")
+		mCalls.IntSum().DataPoints().Append(dpCalls)
+		mCalls.IntSum().SetAggregationTemporality(pdata.AggregationTemporalityCumulative)
+		ilm.Metrics().Append(mCalls)
+	}
+}
+
 // aggregateMetrics aggregates the raw metrics from the input trace data.
 // Each metric is identified by a key that is built from the service name
 // and span metadata such as operation, kind, status_code and any additional
 // dimensions the user has configured.
 func (p *processorImp) aggregateMetrics(traces pdata.Traces) {
-	// TODO: Add implementation
+	for i := 0; i < traces.ResourceSpans().Len(); i++ {
+		rspans := traces.ResourceSpans().At(i)
+		r := rspans.Resource()
+
+		attr, ok := r.Attributes().Get(conventions.AttributeServiceName)
+		if !ok {
+			continue
+		}
+		serviceName := attr.StringVal()
+		p.aggregateMetricsForServiceSpans(rspans, serviceName)
+	}
+}
+
+func (p *processorImp) aggregateMetricsForServiceSpans(rspans pdata.ResourceSpans, serviceName string) {
+	ilsSlice := rspans.InstrumentationLibrarySpans()
+	for j := 0; j < ilsSlice.Len(); j++ {
+		ils := ilsSlice.At(j)
+		spans := ils.Spans()
+		for k := 0; k < spans.Len(); k++ {
+			span := spans.At(k)
+			p.aggregateMetricsForSpan(serviceName, span)
+		}
+	}
+}
+
+func (p *processorImp) aggregateMetricsForSpan(serviceName string, span pdata.Span) {
+	latencyInMilliseconds := float64(span.EndTime()-span.StartTime()) / float64(time.Millisecond.Nanoseconds())
+
+	// Binary search to find the latencyInMilliseconds bucket index.
+	index := sort.SearchFloat64s(p.latencyBounds, latencyInMilliseconds)
+
+	p.lock.Lock()
+	key := p.buildKey(serviceName, span)
+	p.updateCallMetrics(key)
+	p.updateLatencyMetrics(key, latencyInMilliseconds, index)
+	p.lock.Unlock()
+}
+
+// updateCallMetrics increments the call count for the given metric key.
+func (p *processorImp) updateCallMetrics(key string) {
+	p.callSum[key]++
+}
+
+// updateLatencyMetrics increments the histogram counts for the given metric key and bucket index.
+func (p *processorImp) updateLatencyMetrics(key string, latency float64, index int) {
+	if _, ok := p.latencyBucketCounts[key]; !ok {
+		p.latencyBucketCounts[key] = make([]uint64, len(p.latencyBounds))
+	}
+	p.latencySum[key] += latency
+	p.latencyCount[key]++
+	p.latencyBucketCounts[key][index]++
+}
+
+func (p *processorImp) addDimension(key, value string) {
+	// It's worth noting that from pprof benchmarks, WriteString is the most expensive operation of this processor.
+	// Specifically, the need to grow the underlying []byte slice to make room for the appended string.
+	p.metricKeyBuilder.WriteString(value)
+	p.dimensionsBuffer[key] = value
```

Comment:
We do not need to _always_ build `dimensionsBuffer`. We only need to build it if the concatenated key is not found in the cache. Can we delete this line and build the `dimensionsBuffer` only when we discover that the concatenated key is missing in the cache and we need to populate the cache?

Reply:
Okay, I have modified this to only build the map if there is a cache miss of `metricKey`. My implementation involves some duplicated logic due to the slightly different requirements for building a `metricKey` vs. building a the metric key map. Open to suggestions on how to improve this.